pyro/distributions/util.py

```diff
@@ -321,6 +321,20 @@ def matrix_triangular_solve_compat(b, A, upper=True):
         return b.trtrs(A, upper=upper)[0].view(b.size())
 
 
+def log_sum_exp(tensor, scale=1.0, dim=-1):
```

Comment:
The scale arg here seems a little weird. I'd at least move it to the right of `dim`, but consider removing it.

Reply:
The scale argument is adopted from scipy's [log_sum_exp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.misc.logsumexp.html), though using it with tensorized scale was my idea. :) I suppose this will be more numerically stable than using the normalized weights and multiplying will have underflow issues - consider a low weight that is multiplied with a large scale factor.

We also have a `log_sum_exp` in `pyro.util` but using that leads to circular import issues.

Reply:
Ok. Can we move the `pyro.util` version to `pyro.distributions.util`?

Reply:
Sure. I think we can also remove that in a later cleanup if we start using this in `AbstractInfer`, since that is the only place it is used.